core/src/bytes/zarith/zarith.h

```diff
@@ -0,0 +1,162 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+
+#ifndef LEDGER_CORE_ZARITH_H
+#define LEDGER_CORE_ZARITH_H
+
+#include <algorithm>
+#include <vector>
+#include <bytes/BytesReader.h>
+#include <utils/hex.h>
+
+namespace ledger {
+    namespace core {
+        class zarith {
+        public:
+            /*
+             * Most significant bit of each byte is replaced with 1 to let
+             * us know there is still something to read, else 0
+             * To sum up, if the resulting bytes are "chained" together (little endian) and we drop
+             * the most significant bit of each byte, we get the input ("chained") bytes
+             * e.g.
+             * Input:           ad d9                  10101101 11011001  (I)
+             * Output:          d9 db 02      11011001 11011011 00000010
+             * LE:              02 db d9      00000010 11011011 11011001
+             * Drop 1st bit:            (0)0000010 (1)1011011 (1)1011001  == (I)
+            */
+            static std::vector<uint8_t> zSerializeNumber(const std::vector<uint8_t> &inputData) {
+                std::vector<uint8_t> result;
+                size_t id = inputData.size() - 1;
+                uint8_t offset = 0;
+                while (id >= 0) {
+                    auto byte = inputData[id];
+
+                    // We shift current byte
+                    byte <<= offset;
+
+                    // Take shifted bits from previous byte
+                    byte |= id < inputData.size() - 1 ? inputData[id + 1] >> (7 - (offset - 1)) : 0x00;
+
+                    // This boolean will let us know if shifting last byte of inputData
+                    // will generate a new byte (with remaining shifted bytes)
+                    bool isFirst = id == 0;
+                    bool needAdditionalByte = isFirst && inputData[id] >> (7 - offset);
+
+                    // We set first bit to 1 to know that there is a coming byte
+                    if (!isFirst || needAdditionalByte) {
```

Comment:
this will be false only once. Lets always do byte |= 0x80 and then outside the loop `result[result.size()-1] &= 0xEF;`

Reply:
I'm not a huge fan of doing then undoing things ... but what can't we do for optims ?! :trollface: 
By the way, here it's `result[result.size()-1] &= 0x7F;`